src/server/frontend_wayland/basic_surface_event_sink.cpp

```diff
@@ -18,74 +18,124 @@
 
 #include "basic_surface_event_sink.h"
 #include "wl_seat.h"
+#include "wl_pointer.h"
+#include "wl_keyboard.h"
+#include "wl_touch.h"
+#include "wl_surface.h"
+#include "wayland_utils.h"
+
+#include <linux/input-event-codes.h>
 
 namespace mf = mir::frontend;
+namespace geom = mir::geometry;
 
-void mf::BasicSurfaceEventSink::handle_event(EventUPtr&& event)
+mf::BasicSurfaceEventSink::BasicSurfaceEventSink(WlSeat* seat, wl_client* client, WlSurface* surface, WlAbstractMirWindow* window)
+    : seat{seat},
+      client{client},
+      surface{surface},
+      window{window},
+      window_size{geometry::Size{0,0}},
+      destroyed{std::make_shared<bool>(false)}
 {
-    switch (mir_event_get_type(event.get()))
-    {
-    case mir_event_type_resize:
-    {
-        auto* const resize_event = mir_event_get_resize_event(event.get());
-        requested_size = {mir_resize_event_get_width(resize_event), mir_resize_event_get_height(resize_event)};
-        if (requested_size != window_size)
-            send_resize(requested_size);
-        break;
-    }
-    case mir_event_type_input:
-    {
-        auto input_event = mir_event_get_input_event(event.get());
+}
 
-        // Remember the timestamp of any events "signed" with a cookie
-        if (mir_input_event_has_cookie(input_event))
-            timestamp_ns = mir_input_event_get_event_time(input_event);
+mf::BasicSurfaceEventSink::~BasicSurfaceEventSink()
+{
+    *destroyed = true;
+}
 
-        switch (mir_input_event_get_type(input_event))
+void mf::BasicSurfaceEventSink::handle_event(EventUPtr&& event)
+{
+    seat->spawn(run_unless(
+        destroyed,
+        [this, event = std::make_shared<EventUPtr>(move(event))]()
```

Comment:
We want a `std::shared_ptr<MirEvent>` rather than a `std::shared_ptr<std::unique_ptr<MirEvent, ...>>`, right?

And, obviously, we don't even want the `shared_ptr` bit, but pragmatically it's annoying to have move-only lambdas.